sonar-db/src/main/java/org/sonar/db/issue/IssueDao.java

```diff
@@ -47,27 +56,59 @@ public void selectNonClosedIssuesByModule(long componentId, ResultHandler handle
     }
   }
 
-  @CheckForNull
-  public IssueDto selectNullableByKey(DbSession session, String key) {
-    return mapper(session).selectByKey(key);
+  public Optional<IssueDto> selectByKey(DbSession session, String key) {
+    return Optional.fromNullable(mapper(session).selectByKey(key));
   }
 
-  public IssueDto selectByKey(DbSession session, String key) {
-    IssueDto issue = selectNullableByKey(session, key);
-    if (issue == null) {
-      throw new IllegalArgumentException(String.format("Issue key '%s' does not exist", key));
+  public IssueDto selectByKeyOrFail(DbSession session, String key) {
+    Optional<IssueDto> issue = selectByKey(session, key);
+    if (!issue.isPresent()) {
+      throw new RowNotFoundException(String.format("Issue with key [%s] does not exist", key));
     }
-    return issue;
+    return issue.get();
   }
 
-  public List<IssueDto> findByActionPlan(DbSession session, String actionPlan) {
+  public List<IssueDto> selectByActionPlan(DbSession session, String actionPlan) {
     return mapper(session).selectByActionPlan(actionPlan);
   }
 
+  /**
+   * Gets a list issues by their keys. The result does NOT contain {@code null} values for issues not found, so
+   * the size of result may be less than the number of keys. A single issue is returned
+   * if input keys contain multiple occurrences of a key.
+   * <p>Results may be in a different order as input keys (see {@link #selectByOrderedKeys(DbSession, List)}).</p>
+   */
   public List<IssueDto> selectByKeys(DbSession session, List<String> keys) {
     return mapper(session).selectByKeys(keys);
   }
 
+  /**
+   * Gets a list issues by their keys. The result does NOT contain {@code null} values for issues not found, so
+   * the size of result may be less than the number of keys. A single issue is returned
+   * if input keys contain multiple occurrences of a key.
+   * <p>Contrary to {@link #selectByKeys(DbSession, List)}, results are in the same order as input keys.</p>
+   */
+  public Iterable<IssueDto> selectByOrderedKeys(DbSession session, List<String> keys) {
+    List<IssueDto> unordered = selectByKeys(session, keys);
+    return FluentIterable.from(keys).transform(new KeyToIssue(unordered)).filter(Predicates.notNull());
+  }
+
+  private static class KeyToIssue implements Function<String, IssueDto> {
```

Comment:
I don't think as it's not a generic function. It's specific to this method (see the field of type Map). Let's share in a common class the next time it's needed elsewhere.